aten/src/ATen/core/typeid.h

```diff
@@ -81,50 +90,144 @@ AT_DEFINE_HASH_FOR_IDWRAPPER(caffe2::TypeIdentifier)
-               "2. Did you build and install PyTorch and Caffe2 separately? "
-               "For example, this would be the case if you ran scripts/onnx/install.sh or "
-               "scripts/onnx/install-develop.sh prior to Aug 12, 2018 "
-               "(commit 1756daaa7530d).  If so, rebuild using the environment variable "
-               " FULL_CAFFE2=1 (if you build latest master, the ONNX scripts are "
-               "updated to do this for you.) "
-               "For more context, see https://github.com/pytorch/pytorch/issues/10460");
-    }
-    gRegisteredTypeNames().insert(name);
-    gTypeNames()[id] = name;
-#else // __GXX_RTTI
-    if (literal_name.empty()) {
-      gTypeNames()[id] = "(RTTI disabled, cannot show name)";
-    } else {
-      gTypeNames()[id] = literal_name;
-    }
-#endif // __GXX_RTTI
+inline void _CtorNotDefault(void* /*ptr*/, size_t /*n*/) {
+  _ThrowRuntimeTypeLogicError(
+      "Type " + std::string(at::demangle_type<T>()) +
+      " is not default-constructible.");
+}
+
+template <
+    typename T,
+    typename std::enable_if<std::is_default_constructible<T>::value>::type* =
+        nullptr>
+inline TypeMetaData::PlacementNew* _PickCtor() {
+  return &_Ctor<T>;
+}
+
+template <
+    typename T,
+    typename std::enable_if<!std::is_default_constructible<T>::value>::type* =
+        nullptr>
+inline TypeMetaData::PlacementNew* _PickCtor() {
+  return &_CtorNotDefault<T>;
+}
+
+/**
+ * Typed copy function for classes.
+ */
+template <typename T>
+inline void _Copy(const void* src, void* dst, size_t n) {
+  const T* typed_src = static_cast<const T*>(src);
+  T* typed_dst = static_cast<T*>(dst);
+  for (size_t i = 0; i < n; ++i) {
+    typed_dst[i] = typed_src[i];
+  }
+}
+
+/**
+ * A placeholder function for types that do not allow assignment.
+ */
+template <typename T>
+inline void _CopyNotAllowed(
+    const void* /*src*/,
+    void* /*dst*/,
+    size_t /*n*/) {
+  _ThrowRuntimeTypeLogicError(
+      "Type " + std::string(at::demangle_type<T>()) +
+      " does not allow assignment.");
+}
+
+template <
+    typename T,
+    typename std::enable_if<std::is_copy_assignable<T>::value>::type* =
+        nullptr>
+inline TypeMetaData::TypedCopy* _PickCopy() {
+  return &_Copy<T>;
+}
+
+template <
+    typename T,
+    typename std::enable_if<!std::is_copy_assignable<T>::value>::type* =
+        nullptr>
+inline TypeMetaData::TypedCopy* _PickCopy() {
+  return &_CopyNotAllowed<T>;
+}
+
+/**
+ * Destructor for non-fundamental types.
+ */
+template <typename T>
+inline void _Dtor(void* ptr, size_t n) {
+  T* typed_ptr = static_cast<T*>(ptr);
+  for (size_t i = 0; i < n; ++i) {
+    typed_ptr[i].~T();
   }
+}
+
+template<class T> const char* _TypeName() noexcept;
+
+template<class T, class Enable = void>
+struct TypeMetaDataRegistry final {
+  static_assert(!std::is_same<T, T>::value, "This should never be picked since TypeMetaDataRegistry has specialisations for all cases");
```

Comment:
Err... why can't you just say `false` here?

Reply:
because static_asserts that don't depend on a type are evaluated even if the class isn't instantiated, i.e. a static_assert(false) immediately errors. Using this false that depends on `T`, it will only faild if the class is actually instantiated.

Reply:
oh right, forgot about that! (I feel ashamed, because I just a few days ago used this trick in one of my own patches haha)